Conditional endorsement transformations (#299)
ietf-rats-wg/draft-ietf-rats-corim: draft-ietf-rats-corim.md

#### Reference Values Triple {#sec-comid-triple-refval}

The Reference Values Triple has the following structure:

~~~ cddl
{::include cddl/reference-triple-record.cddl}
~~~

The `reference-triple-record` has the following parameters:

> * `ref-env`: Search criteria that target Evidence environments in the ACS.

Reviewer: This seems to be a significant shift in the earlier description of the triple! The triple semantics should not require us to describe where the triple is targeted to internally, in the description of the triple.

It should simply say:

What this triple is about :
What does it convey
i.e it conveys for a target Env identified by the Env: The Reference measurements identified by ref-claims

it should not expain anything about the ACS here

Reviewer 2: Separate comment: I am not sure I know how to parse that sentence. Quick help: what's an "Evidence environment"? I assume I am missing something obvious here.

Author: > Separate comment: I am not sure I know how to parse that sentence. Quick help: what's an "Evidence environment"? I assume I am missing something obvious here.

Is it more clear to just say: "Search criteria that targets Evidence."

> * `ref-env`: Search criteria that target Evidence environments in the ACS.

Reviewer 2: Would this be still correct?
```suggestion
* `ref-env`: Search criteria that select certain Evidence Claims in the ACS.
```

* `ref-claims`: Search criteria that target Evidence Claims in the ACS.

To process `reference-triple-record` both the `ref-env` and `ref-claims` are compared with evidence entries from the ACS.
If these conditions are met, the matching ACS entry is copied into a new ACS entry that is added to the ACS, but with the Reference Value Provider's authority.

#### Endorsed Values Triple {#sec-comid-triple-endval}

The Endorsed Values Triple has the following structure:

~~~ cddl
{::include cddl/endorsed-triple-record.cddl}
~~~

The `endorsed-triple-record` has the following parameters:

> * `condition`: Search criteria applied to the ACS that determine if the `endorsement` applies.

Reviewer: This section should say:
--> What triple is intended to convey

Versus, 

What CoRIM spec is guiding users to use a possible internal representation in terms of ACS.

This document should be very much applicable and usable to some Supply Chain Consumers, who want to use the Triple based on their semantics but do NOT use Verifier logic which uses ACS but have another means to process the Triple and Implement a Verifier.

I am little concerned that we have moved ACS and internal representation far too much into the base triple description.

Author: Is it more clear to say: "Search criteria that targets Attester state."

Reviewer 3: I think we should not shy away from the mechanics here, i.e., linking triples and ACS, because it is necessary for the reader to understand the connection between the two. However, it makes sense to transition from abstract to practical in a smoother way.

Reviewer: @thomas-fossati : I tend to dis-agree: I see the clear distinction as below:

Triple should describe what it is meant for : For Ref Triple, used to convey Ref values
for the Env in subject with Reference Measurements/claims in the object. These measurements are used by Verifier to match against the same measurements as received from Attester in Evidence. One possible way of converting these measurements internally and matching logics are described in section : XYZ:

_In another section (XYZ outside this section)_ 
How a potential one use of the triple by converting into a representation best use for matching is where ACS comes into picture!

Reviewer: > Is it more clear to say: "Search criteria that targets Attester state."

Yes, at high level, it is fine to say something about matching with Attester Evidence claims

#### Reference Values Triple {#sec-comid-triple-refval}

The Reference Values Triple has the following structure:

~~~ cddl
{::include cddl/reference-triple-record.cddl}
~~~

> The `reference-triple-record` has the following parameters:

Reviewer: Much better text than earlier, however I would add your text on top of the old text purely for simplicity and logical enhancement. I suggest please re-consider this way.

> The `reference-triple-record` has the following parameters:
> 
> * `ref-env`: Search criterion that locates an Evidence environment that matches the reference environment.
> * `ref-claims`: Search criteria that locates the Evidence measurements that match the reference Claims.
> 
> To process `reference-triple-record` both the `ref-env` and `ref-claims` criteria are compared with Evidence entries.
> If the search criteria are satisfied, the matching entry is re-asserted, except with the Reference Value Provider's authority.
> By re-asserting Evidence using the RVP's authority, the Verifier can avoid mixing Reference Values (reference state) with Evidence (actual state).
> See {{-rats-endorsements}}.
> Re-asserted Evidence using RVP authority is said to be "corroborated".
> 

Reviewer: ```suggestion
A Reference Values triple relates reference measurements to a Target
Environment. For Reference Value Claims, the subject identifies a Target
Environment, i.e ref-env, the object contains measurements, i.e. ref-claims, and the predicate asserts that
these are the expected (i.e., reference) measurements for the Target Environment.

Within a Verifier, they are used as under:

`ref-env`: Search criterion to locate an Evidence environment that matches the reference environment.
* `ref-claims`: Search criteria to locate the Evidence measurements that match the reference Claims.

To process `reference-triple-record` both the `ref-env` and `ref-claims` criteria are compared with Evidence entries.
If the search criteria are satisfied, the matching entry is re-asserted, except with the Reference Value Provider's authority.
By re-asserting Evidence using the RVP's authority, the Verifier can avoid mixing Reference Values (reference state) with Evidence (actual state).
See {{-rats-endorsements}}.
Re-asserted Evidence using RVP authority is said to be "corroborated".

```

Reviewer: Raised issue #310 to address the issue and concerns I have!

#### Endorsed Values Triple {#sec-comid-triple-endval}

The Endorsed Values Triple has the following structure:

~~~ cddl
{::include cddl/endorsed-triple-record.cddl}
~~~

> The `endorsed-triple-record` has the following parameters:

Reviewer: ```suggestion

The `endorsed-triple-record` conveys additional Endorsements claims for a Target Environment   to a Verifier.

The `endorsed-triple-record` has the following parameters:
```

Reviewer: Raised issue #310 to address the concern I have!

> * `condition`: Search criterion that locates Evidence, corroborated Evidence, or Endorsements.

Reviewer 3: envidence [etc] _environment_, if we want to keep this consistent with RVT above

Author: > envidence [etc] _environment_, if we want to keep this consistent with RVT above

Addressed by commit https://github.com/ietf-rats-wg/draft-ietf-rats-corim/pull/299/commits/31bc933253de02567f3c9dc184771be189436f6e

#### Reference Values Triple {#sec-comid-triple-refval}

The Reference Values Triple has the following structure:

~~~ cddl
{::include cddl/reference-triple-record.cddl}
~~~

The `reference-triple-record` has the following parameters:

* `ref-env`: Search criteria that targets Evidence environments in the ACS.
* `ref-claims`: Search criteria that targets Evidence Claims in the ACS.

> To process `reference-triple-record` both the `ref-env` and `ref-claims` are compared with entries from the ACS.

Reviewer: ```suggestion
To process `reference-triple-record` both the `ref-env` and `ref-claims` are compared with evidence entries from the ACS.
```

If these conditions are met, the matching ACS entry is copied into a new ACS entry that is added to the ACS,
but with the Reference Value Provider's authority.

#### Endorsed Values Triple {#sec-comid-triple-endval}

The Endorsed Values Triple has the following structure:

~~~ cddl
{::include cddl/endorsed-triple-record.cddl}
~~~

The `endorsed-triple-record` has the following parameters:

> * `condition`: Search criteria applied to the ACS that dermines if the `endorsement` applies.

Reviewer 4: ```suggestion
* `condition`: Search criteria applied to the ACS that determines if the `endorsement` applies.
```

* `endorsement`: Claims to be added to the ACS.

> To proess a `endorsed-triple-record` the `condition` is compared with entries from the ACS.

Reviewer 4: ```suggestion
To process a `endorsed-triple-record` the `condition` is compared with entries from the ACS.
```

If the condition is met, the `endorsement` is combined with the `environment-map` of the matched `condition` to form a new ACS entry.
The new ACS entry is added to the ACS with the Endorser's authority.

#### Conditional Endorsement Triple {#sec-comid-triple-cond-endors}

The Conditional Endorsement Triple has the following structure:

~~~ cddl
{::include cddl/conditional-endorsement-triple-record.cddl}
~~~

~~~ cddl
{::include cddl/stateful-environment-record.cddl}
~~~

The `conditional-endorsement-triple-record` has the following parameters:

> * `conditions`: Search criteria applied to the ACS that dermines if the `endorsements` apply.

Reviewer 4: ```suggestion
* `conditions`: Search criteria applied to the ACS that determines if the `endorsements` apply.
```

* `endorsements`: Sets of Claims to be added to the ACS.

To process a `conditional-endorsement-triple-record` the `conditions` are compared with entries from the ACS.
If all the conditions are met, the `endorsements` are added to the ACS with the Endorser's authority.

#### Conditional Endorsement Series Triple {#sec-comid-triple-cond-series}

The Conditional Endorsement Series Triple has the following structure:

~~~ cddl
{::include cddl/conditional-endorsement-series-triple-record.cddl}
~~~

~~~ cddl
{::include cddl/conditional-series-record.cddl}
~~~

The `conditional-endorsement-series-triple-record` has the following parameters:

> * `condition`: Search criteria applied to the ACS that dermines if the `series` applies.

Reviewer 4: ```suggestion
* `condition`: Search criteria applied to the ACS that determines if the `series` applies.
```

* `series`: A set of selection-addition tuples.

The `conditional-series-record` has the following parameters:

* `selection`: Search criteria applied to the ACS entries resulting from the matched `conditions`
* `addition`: Claims to be added to the ACS is the `selection` is satisfied.

> To process a`conditional-endorsement-series-record` the `conditions` are compared with entries from the ACS.

Reviewer 4: ```suggestion
To process a `conditional-endorsement-series-record`, first the `condition` is compared with entries from the ACS.
```

Author: All of the conditions could be compared unless one of them matches.

Author: whitespace fix applied

{::include cddl/endorsed-triple-record.cddl}
~~~

#### Conditional Endorsement Triple {#sec-comid-triple-cond-endors}

> The semantics of the Conditional Endorsement Triple is as follows:

Reviewer: After commuting my change, PR #276, I realised we need to work on Conditional Endorsement Triple.

The subject is a Series of `stateful-environment-record` 

However the predicate should be a series of `measurement-map` NOT a series of Endorsement triples, as the Environment simply does not make sense in the predicate,

Author: See revised text.

> "IF accepted state matches all `conds` values, THEN every entry in the `endorsements` is added to the accepted state"

~~~ cddl
{::include cddl/conditional-endorsement-triple-record.cddl}
~~~

A `conditional-endorsement-triple-record` has the following parameters:

* `conditions`: all target environments, along with a specific state, that need to match `state-triples` entries in the ACS for the endorsement(s) to apply
* `endorsements`: endorsements that are added to the ACS `state-triples` if all `conds` match.

The order in which Conditional Endorsement triples are evaluated is important: different sorting may produce different end-results in the computed ACS.

> Therefore, the set of applicable Conditional Endorsement triples MUST be topologically sorted based on the criterion that a Conditional Endorsement triple is evaluated before another if its Target Environment and Endorsement pair is found in any of the stateful environments of the subsequent triple.

Reviewer 4: I don't think topological sorting is going to be a complete solution as profiles get more and more complex. We should be computing a fixed point. The ordering of conditions may not be easily decided, but if it is, then we can suggest a topological sort to improve convergence speed.

Author: See revised text.

> * The `element-list` field is populated with the measurements collected by an Attesting Environment.
> * The `authority` field is populated with the identity of the entity that asserted (e.g., signed) the Conceptual Message.
> * The `cmtype` field is set based on the type of Conceptual Message inputted or to be outputed.
> * The `profile` field is set based on the `corim-map` `profile` value.

#### Appraisal Context Construction

All of the extracted and validated tags are loaded into an *appraisal context*.
The Appraisal Context contains an internal representation of the inputted Conceptual Messages.
The selected tags are mapped to an internal representation, making them suitable for appraisal processing.

#### Reference Triples Transformation {#sec-ref-trans}

{:rtt-enum: counter="foo" style="format Step %d."}

{: rtt-enum}
* An available `rv` list entry ({{sec-ir-ref-val}}) is allocated.

* The ECT `cmtype` of the `ev` entry is set to `reference-values`.

* The Reference Values Triple ({{sec-comid-triple-refval}}) `environment-map` is copied to the `environment-map` for both the `rv` `condition` and `rv` `addition` ECTs.

> * For each `measurement-map` entry in the measurements list, the i<sup>th</sup> `measurement-map` is copied to the i<sup>th</sup> `element-map` in the `element-list` of the `rv` `addition` ECT.

Reviewer 4: I didn't think order would matter, given that mkey is the distinguishing feature, and not the position in the element-list.

Author: > I didn't think order would matter, given that mkey is the distinguishing feature, and not the position in the element-list.

The processing sections pay attention to mkey semantics. This is saying that the transformation isn't expected to change position semantics. It may be harmless but it's not my intention to do that here.

There is a yet-to-be-discussed issue regarding what the processing requirements are if mkey is omitted and multiples of measurement-map exist. The CDDL permits this. I think it's reasonable for the verifier use array position to disambiguate. The RIM author decides which approach is better. Otherwise, we have to put normative next banning this condition.

Author: Issue #303 created and tracking annotations added.

* If the Endorsement conceptual message has a profile, the profile is copied to the `ev` `addition` ECT `profile` field.

#### Endorsement Triples Transformations {#sec-end-trans}

Endorsement Triple Transformation :

{:ett-enum: counter="ett" style="format Step %d."}

{: ett-enum}
* An available `ev` entry ({{sec-ir-end-val}}) is allocated.

* The ECT `cmtype` of the `ev` entry's `addition` ECT is set to `endorsements`.

* The Endorsement Triple ({{sec-comid-triple-endval}}) `environment-map` is copied to the `environment-map` for both the `ev` `condition` and `ev` `addition` ECTs.

* For each `measurement-map` entry in the measurements list, the i<sup>th</sup> `measurement-map` entry is copied to the i<sup>th</sup> entry in the `addition` ECT `element-map` of the `element-list`.

* The issuer of the Endorsement conceptual message is copied to the `ev` `addition` ECT `authority` field.

* If the Endorsement conceptual message has a profile, the profile is copied to the `ev` `addition` ECT `profile` field.

Conditional Endorsement Triple Transformation :

{:cett-enum: counter="cett" style="format Step %d."}

{: cett-enum}

> * An available `ev` entry ({{sec-ir-end-val}}) is allocated.

Reviewer 4: I know the words aren't new in this PR, but what does it mean for the `ev` entry to be "available"? I don't think we need to talk about allocation behavior. Are we not simply populating the appraisal context with this triple as one to be processed against the ACS?

Author: Maybe it's ok to remove it?

### Input Transformation {#sec-phase1-trans}

Input Conceptual Messages, whether Endorsements, Reference Values, Evidence, or Policies, are transformed to an internal representation that is based on ECTs ({{sec-ir-cm}}).

The following mapping conventions apply to all forms of input transformation:

> * The `environment` field is populated with a Target Environment identifier.
> * The `element-list` field is populated with the measurements collected by an Attesting Environment.
> * The `authority` field is populated with the identity of the entity that asserted (e.g., signed) the Conceptual Message.

> > * The `cmtype` field is set based on the type of Conceptual Message inputted or to be outputed.

Reviewer 3: ```suggestion
> * The `cmtype` field is set based on the type of Conceptual Message inputted or to be output.
```

> > * The `profile` field is set based on the `corim-map` `profile` value.

Reviewer 4: Why put in a quote block when that's not what we do elsewhere?

Author: > Why put in a quote block when that's not what we do elsewhere?

That is what was there already. The previous section also uses similar formatting. BTW It isn't a quote block it is markdown for indentation.

#### Appraisal Context Construction

All of the extracted and validated tags are loaded into an *appraisal context*.
The Appraisal Context contains an internal representation of the inputted Conceptual Messages.
The selected tags are mapped to an internal representation, making them suitable for appraisal processing.

#### Reference Triples Transformation {#sec-ref-trans}

{:rtt-enum: counter="foo" style="format Step %d."}

{: rtt-enum}

> * An `rv` list entry ({{sec-ir-ref-val}}) is allocated.

Reviewer 4: If we allow reference value triples to apply to endorsed measurements, then the `rv` relation is the same as the `ev` relation. I would say we don't need it.

Author: CoRIM is written from the perspective that is laid out by RFC9334 where the roles of Reference Value Provider and Endorser differ. The interaction between RVPs, Attesters and RPs is meaningful because of the conceptual messages they exchange. There isn't the notion of a role credential. The result of an RVP-Attester interaction is "corroborated" Ref Values that have cmtype=`reference` but is an "actual state" representation in the ACS. If an Endorser constructed an endorsement rule that produced a record with the same claims as a corroborated RV contains, it would still be an Endorsement. RFC9334 expects RPs will differentiate between the roles and will be looking for statements from RVPs regarding acceptability of Evidence vs. a similar construction from Endorsers.

Reviewer 4: I understand that the roles are different, but the behavior of `rv` as an internal representation can easily be expressed as `ev` since you're already populating the additions with the appropriate `cmtype`. That's all I was saying.

Author: > I understand that the roles are different, but the behavior of `rv` as an internal representation can easily be expressed as `ev` since you're already populating the additions with the appropriate `cmtype`. That's all I was saying.

The internal representation is primarily there to facilitate readability. Having distinct representations for the various conceptual message types keeps the text inside the conceptual message swim lanes. It's a style thing more than an optimization thing.

* The issuer of the Endorsement conceptual message is copied to the `ev` `addition` ECT `authority` field.

* If the Endorsement conceptual message has a profile, the profile is copied to the `ev` `addition` ECT `profile` field.

Conditional Endorsement Triple Transformation :

{:cett-enum: counter="cett" style="format Step %d."}

{: cett-enum}
* An available `ev` entry ({{sec-ir-end-val}}) is allocated.

* The ECT `cmtype` of the `ev` entry's `addition` ECT is set to `endorsements`.

* For each entry in the Conditional Endorsement Triple ({{sec-comid-triple-cond-endors}}) `conditions` array, the `stateful-environment-record` is copied to a `condition` ECT in the `ev` entry.
The `environment-map` of the `stateful-environment-record` is copied to the ECT `environment` field.

* For each `measurement-map` entry in the measurements list of the `stateful-environment-record`, the i<sup>th</sup> `measurement-map` entry is copied to the i<sup>th</sup> entry in the `condition` ECT `element-map` of the `element-list`.

* For each entry in the Conditional Endorsement Triple `endorsements` array, the `endorsed-triple-record` is copied to an `addition` ECT in the `ev` entry.

> The `environment-map` of the `endorsed-triple-record` is copied to the ECT `environment` field.

Reviewer: i think the environment map of the `endorsed-triple-record` is the environment to which the Endorsement applies.

It should reside, where the addition ECT, i.e. place where the stuff gets added as Endorsement exist

Reviewer 4: I'm not seeing where the issue is. The text seems fine to me since "the ECT" refers to the addition ECT that corresponds to an entry in the conditional endorsement triple's `endorsements` array.

#### Appraisal Context Construction

All of the extracted and validated tags are loaded into an *appraisal context*.
The Appraisal Context contains an internal representation of the inputted Conceptual Messages.
The selected tags are mapped to an internal representation, making them suitable for appraisal processing.

#### Reference Triples Transformation {#sec-ref-trans}

{:rtt-enum: counter="foo" style="format Step %d."}

{: rtt-enum}
* An `rv` list entry ({{sec-ir-ref-val}}) is allocated.

* The `cmtype` of the `addition` ECT in the `rv` entry is set to `reference-values`.

* The Reference Values Triple ({{sec-comid-triple-refval}}) `environment-map` is copied to the `environment-map` for both the `rv` `condition` and `rv` `addition` ECTs.

* For each `measurement-map` entry in the measurements list, the i<sup>th</sup> `measurement-map` is copied to the i<sup>th</sup> `element-map` in the `element-list` of the `rv` `addition` ECT.

[^issue] https://github.com/ietf-rats-wg/draft-ietf-rats-corim/issues/303

* The issuer of the Endorsement conceptual message is copied to the `ev` `addition` ECT `authority` field.

> * If the Endorsement conceptual message has a profile, the profile is copied to the `ev` `addition` ECT `profile` field.

Author: ```suggestion
* If the Endorsement conceptual message has a profile, the profile identifier is copied to the `ev` `addition` ECT `profile` field.
```

If satisfied, the RVP authority is added to the matching ACS entry.

Reference Values are matched with ACS entries by iterating through the `rv` list.

> For each `rv` entry, the `condition` ECT is compared with an ACS ECT, where the ACS ECT `cmtype` contains `evidence`.

Reviewer 4: Can one party's reference values not be for another party's endorsements?

Author: > Can one party's reference values not be for another party's endorsements?

Structurally, that would be reasonable since both Evidence and Endorsements are expressions of actual state. However, I don't know why it is necessary to model a condition on endorsements using RVs since a conditional endorsement condition could be expressed using possible-state expressions. I'll review the current wording, but I believe the cmtype usage is specific to the addition ECTs and silent on the condition ECTs.

Author: Added an issue and tracking annotation.

If the ECTs match except for authority, the `rv` `addition` ECT authority is added to the ACS ECT authority.

## Endorsed Values Augmentation (Phase 4) {#sec-phase4}

[^issue] https://github.com/ietf-rats-wg/draft-ietf-rats-corim/issues/179

Endorsers publish Endorsements using endorsement triples (see {{sec-comid-triple-endval}, {{sec-comid-triple-cond-endors}}, and {{sec-comid-triple-cond-series}}) which are transformed ({{sec-end-trans}}) into an internal representation ({{sec-ir-end-val}}).
Endorsements describe actual Attester state.
Endorsements are added to the ACS if the Endorsement condition is satisifed by the ACS.

### Processing Endorsements {#sec-process-end}

> Endorsements are matched with ACS entries by iterating through the `ev` list.

Reviewer 4: The "`ev` list" I think needs to be more explicitly understood as the collection in the appraisal context I think. Same for `evs` below?

For each `ev` entry, the `condition` ECT is compared with an ACS ECT, where the ACS ECT `cmtype` contains either `evidence` or `endorsements`.
If the ECTs match ({{sec-match-condition-ect}}), the `ev` `addition` ECT is added to the ACS.

### Processing Conditional Endorsements {#sec-process-cond-end}

Conditional Endorsement Triples are transformed into an internal representation based on `ev`.
Conditional endorsements have the same processing steps as shown in ({{sec-process-end}}).

### Processing Conditional Endorsement Series {#sec-process-series}

Conditional Endorsement Series Triples are transformed into an internal representation based on `evs`.
Conditional series endorsements are matched with ACS entries first by iterating through the `evs` list,

> where for each `ev` entry, the `condition` ECT is compared with an ACS ECT, where the ACS ECT `cmtype` contains either `evidence` or `endorsements`.

Reviewer 4: Iterating through the entries with `cmtype` is `endorsement`? There's no relation entries of the ACS, and there's no ev in an evs. This sentence confuses me.

Author: It looks like a typo.

##### Comparison for raw-value entries

> > [^issue] *I think this comparison method only works if the entry is at key 4 (because
> there needs to be a mask at key 5). Should we have a Reference Value of this
> which stores `[expect-raw-value raw-value-mask]` in an array?*

Reviewer 3: This line can be removed.  The section is still empty, it's unclear what this text refers to.

```suggestion
```

: binary

Security considerations:

> : {{sec-sec}}) of {{&SELF}}

Reviewer 3: ```suggestion
: {{sec-sec}} of {{&SELF}}
```
